// angular2-json-schema-form/src/frameworks/material-design/material-checkbox.component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Component, Input } from '@angular/core';
import { JsonSchemaFormService } from '../../library/json-schema-form.service';
var MaterialCheckboxComponent = (function () {
    function MaterialCheckboxComponent(jsf) {
        this.jsf = jsf;
        this.controlDisabled = false;
        this.boundControl = false;
        this.trueValue = true;
        this.falseValue = false;
    }
    MaterialCheckboxComponent.prototype.ngOnInit = function () {
        this.options = this.layoutNode.options;
        this.jsf.initializeControl(this);
        if (this.controlValue === null || this.controlValue === undefined) {
            this.controlValue = this.options.title;
        }
    };
    MaterialCheckboxComponent.prototype.updateValue = function (event) {
        this.jsf.updateValue(this, event.checked ? this.trueValue : this.falseValue);
    };
    return MaterialCheckboxComponent;
}());
__decorate([
    Input(),
    __metadata("design:type", Number)
], MaterialCheckboxComponent.prototype, "formID", void 0);
__decorate([
    Input(),
    __metadata("design:type", Object)
], MaterialCheckboxComponent.prototype, "layoutNode", void 0);
__decorate([
    Input(),
    __metadata("design:type", Array)
], MaterialCheckboxComponent.prototype, "layoutIndex", void 0);
__decorate([
    Input(),
    __metadata("design:type", Array)
], MaterialCheckboxComponent.prototype, "dataIndex", void 0);
MaterialCheckboxComponent = __decorate([
    Component({
        selector: 'material-checkbox-widget',
        template: "\n    <md-checkbox\n      align=\"left\"\n      [color]=\"options?.color || 'accent'\"\n      [disabled]=\"controlDisabled || options?.readonly\"\n      [id]=\"'control' + layoutNode?._id\"\n      [name]=\"controlName\"\n      [checked]=\"controlValue\"\n      (change)=\"updateValue($event)\">\n      <span *ngIf=\"options?.title\"\n        [class.sr-only]=\"options?.notitle\"\n        [innerHTML]=\"options?.title\"></span>\n    </md-checkbox>",
    }),
    __metadata("design:paramtypes", [JsonSchemaFormService])
], MaterialCheckboxComponent);
export { MaterialCheckboxComponent };
