// angular2-json-schema-form/src/frameworks/material-design/material-card.component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Component, Input } from '@angular/core';
import { JsonSchemaFormService } from '../../library/json-schema-form.service';
var MaterialCardComponent = (function () {
    function MaterialCardComponent(jsf) {
        this.jsf = jsf;
        this.expanded = true;
    }
    MaterialCardComponent.prototype.ngOnInit = function () {
        this.options = this.layoutNode.options;
        this.expanded = !this.options.expandable;
    };
    MaterialCardComponent.prototype.expand = function () {
        if (this.options.expandable)
            this.expanded = !this.expanded;
    };
    return MaterialCardComponent;
}());
__decorate([
    Input(),
    __metadata("design:type", Number)
], MaterialCardComponent.prototype, "formID", void 0);
__decorate([
    Input(),
    __metadata("design:type", Object)
], MaterialCardComponent.prototype, "layoutNode", void 0);
__decorate([
    Input(),
    __metadata("design:type", Array)
], MaterialCardComponent.prototype, "layoutIndex", void 0);
__decorate([
    Input(),
    __metadata("design:type", Array)
], MaterialCardComponent.prototype, "dataIndex", void 0);
MaterialCardComponent = __decorate([
    Component({
        selector: 'material-card-widget',
        template: "\n    <md-card\n      [attr.disabled]=\"options?.readonly\"\n      [class]=\"options?.htmlClass\"\n      [class.expandable]=\"options?.expandable && !expanded\"\n      [class.expanded]=\"options?.expandable && expanded\">\n      <md-card-title-group *ngIf=\"options?.title || options?.description\"\n        [class]=\"options?.labelHtmlClass\"\n        (click)=\"expand()\">\n        <md-card-title *ngIf=\"options?.title\"\n          [class.sr-only]=\"options?.notitle\"\n          [innerHTML]=\"options?.title\">{{options?.title}}</md-card-title>\n        <md-card-subtitle *ngIf=\"options?.description\">{{options?.description}}</md-card-subtitle>\n      </md-card-title-group>\n      <md-card-content>\n        <root-widget *ngIf=\"expanded\"\n          [formID]=\"formID\"\n          [layout]=\"layoutNode.items\"\n          [dataIndex]=\"dataIndex\"\n          [layoutIndex]=\"layoutIndex\"\n          [isOrderable]=\"options?.orderable\"></root-widget>\n      </md-card-content>\n    </md-card>",
        styles: ["\n    .expandable > legend:before { content: '\\25B8'; padding-right: .3em; }\n    .expanded > legend:before { content: '\\25BE'; padding-right: .2em; }\n  "],
    }),
    __metadata("design:paramtypes", [JsonSchemaFormService])
], MaterialCardComponent);
export { MaterialCardComponent };
