// angular2-json-schema-form/src/frameworks/material-design/material-textarea.component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Component, Input } from '@angular/core';
import { JsonSchemaFormService } from '../../library/json-schema-form.service';
var MaterialTextareaComponent = (function () {
    function MaterialTextareaComponent(jsf) {
        this.jsf = jsf;
        this.controlDisabled = false;
        this.boundControl = false;
    }
    MaterialTextareaComponent.prototype.ngOnInit = function () {
        this.options = this.layoutNode.options;
        this.jsf.initializeControl(this);
    };
    MaterialTextareaComponent.prototype.updateValue = function (event) {
        this.jsf.updateValue(this, event.target.value);
    };
    return MaterialTextareaComponent;
}());
__decorate([
    Input(),
    __metadata("design:type", Number)
], MaterialTextareaComponent.prototype, "formID", void 0);
__decorate([
    Input(),
    __metadata("design:type", Object)
], MaterialTextareaComponent.prototype, "layoutNode", void 0);
__decorate([
    Input(),
    __metadata("design:type", Array)
], MaterialTextareaComponent.prototype, "layoutIndex", void 0);
__decorate([
    Input(),
    __metadata("design:type", Array)
], MaterialTextareaComponent.prototype, "dataIndex", void 0);
MaterialTextareaComponent = __decorate([
    Component({
        selector: 'material-textarea-widget',
        template: "\n    <md-textarea #inputControl\n      [attr.aria-describedby]=\"'control' + layoutNode?._id + 'Status'\"\n      [attr.list]=\"'control' + layoutNode?._id + 'Autocomplete'\"\n      [attr.maxlength]=\"options?.maxLength\"\n      [attr.minlength]=\"options?.minLength\"\n      [attr.pattern]=\"options?.pattern\"\n      [attr.required]=\"options?.required\"\n      [class]=\"options?.fieldHtmlClass\"\n      [disabled]=\"controlDisabled\"\n      [id]=\"'control' + layoutNode?._id\"\n      [name]=\"controlName\"\n      [placeholder]=\"options?.title\"\n      [readonly]=\"options?.readonly ? 'readonly' : null\"\n      [style.width]=\"'100%'\"\n      [value]=\"controlValue\"\n      (input)=\"updateValue($event)\">\n      <span *ngIf=\"options?.fieldAddonLeft\"\n        md-prefix>{{options?.fieldAddonLeft}}</span>\n      <span *ngIf=\"options?.fieldAddonRight\"\n        md-suffix>{{options?.fieldAddonRight}}</span>\n      <md-hint *ngIf=\"options?.description\"\n        align=\"end\">{{options?.description}}</md-hint>\n      <md-hint *ngIf=\"options?.placeholder && !formControl?.dirty\"\n        align=\"end\">{{options?.placeholder}}</md-hint>\n    </md-textarea>",
    }),
    __metadata("design:paramtypes", [JsonSchemaFormService])
], MaterialTextareaComponent);
export { MaterialTextareaComponent };
