// angular2-json-schema-form/src/library/json-schema-form.module.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { NgModule, ANALYZE_FOR_ENTRY_COMPONENTS } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { MaterialModule } from '@angular/material';
import { JsonSchemaFormComponent } from './json-schema-form.component';
import { OrderableDirective } from './utilities/orderable.directive';
import { ALL_FRAMEWORKS } from '../frameworks/index';
import { ALL_WIDGETS } from '../widgets/index';
import { ALL_MATERIAL_DESIGN_WIDGETS } from '../frameworks/material-design/index';
import { FrameworkLibraryService } from '../frameworks/framework-library.service';
import { WidgetLibraryService } from '../widgets/widget-library.service';
import { JsonSchemaFormService } from './json-schema-form.service';
import 'ajv';
import 'lodash';
export { FrameworkLibraryService, WidgetLibraryService, JsonSchemaFormService, JsonSchemaFormComponent, };
var ALL_COMPONENTS = ALL_FRAMEWORKS.concat(ALL_WIDGETS, ALL_MATERIAL_DESIGN_WIDGETS);
var JsonSchemaFormModule = JsonSchemaFormModule_1 = (function () {
    function JsonSchemaFormModule() {
    }
    JsonSchemaFormModule.forRoot = function () {
        return {
            ngModule: JsonSchemaFormModule_1,
            providers: [
                FrameworkLibraryService,
                WidgetLibraryService,
                {
                    provide: ANALYZE_FOR_ENTRY_COMPONENTS,
                    useValue: ALL_COMPONENTS.slice(),
                    multi: true,
                },
            ],
        };
    };
    return JsonSchemaFormModule;
}());
JsonSchemaFormModule = JsonSchemaFormModule_1 = __decorate([
    NgModule({
        imports: [
            CommonModule, FormsModule, ReactiveFormsModule, MaterialModule,
        ],
        declarations: [
            JsonSchemaFormComponent, OrderableDirective
        ].concat(ALL_COMPONENTS),
        exports: [
            FormsModule, ReactiveFormsModule, MaterialModule,
            JsonSchemaFormComponent, OrderableDirective,
        ],
    })
], JsonSchemaFormModule);
export { JsonSchemaFormModule };
var JsonSchemaFormModule_1;
