// angular2-json-schema-form/src/widgets/message.component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Component, Input } from '@angular/core';
import { JsonSchemaFormService } from '../library/json-schema-form.service';
var MessageComponent = (function () {
    function MessageComponent(jsf) {
        this.jsf = jsf;
        this.message = null;
    }
    MessageComponent.prototype.ngOnInit = function () {
        this.options = this.layoutNode.options;
        this.message = this.options.help || this.options.helpvalue ||
            this.options.msg || this.options.message;
    };
    return MessageComponent;
}());
__decorate([
    Input(),
    __metadata("design:type", Number)
], MessageComponent.prototype, "formID", void 0);
__decorate([
    Input(),
    __metadata("design:type", Object)
], MessageComponent.prototype, "layoutNode", void 0);
__decorate([
    Input(),
    __metadata("design:type", Array)
], MessageComponent.prototype, "layoutIndex", void 0);
__decorate([
    Input(),
    __metadata("design:type", Array)
], MessageComponent.prototype, "dataIndex", void 0);
MessageComponent = __decorate([
    Component({
        selector: 'message-widget',
        template: "\n    <span *ngIf=\"message\"\n      [class]=\"options?.labelHtmlClass\"\n      [innerHTML]=\"message\"></span>",
    }),
    __metadata("design:paramtypes", [JsonSchemaFormService])
], MessageComponent);
export { MessageComponent };
