// angular2-json-schema-form/src/widgets/tab.component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Component, Input } from '@angular/core';
import { JsonSchemaFormService } from '../library/json-schema-form.service';
var TabComponent = (function () {
    function TabComponent(jsf) {
        this.jsf = jsf;
    }
    TabComponent.prototype.ngOnInit = function () {
        this.options = this.layoutNode.options;
    };
    return TabComponent;
}());
__decorate([
    Input(),
    __metadata("design:type", Number)
], TabComponent.prototype, "formID", void 0);
__decorate([
    Input(),
    __metadata("design:type", Object)
], TabComponent.prototype, "layoutNode", void 0);
__decorate([
    Input(),
    __metadata("design:type", Array)
], TabComponent.prototype, "layoutIndex", void 0);
__decorate([
    Input(),
    __metadata("design:type", Array)
], TabComponent.prototype, "dataIndex", void 0);
TabComponent = __decorate([
    Component({
        selector: 'tab-widget',
        template: "\n    <div\n      [class]=\"options?.htmlClass\">\n      <root-widget\n        [formID]=\"formID\"\n        [layout]=\"layoutNode.items\"\n        [dataIndex]=\"dataIndex\"\n        [layoutIndex]=\"layoutIndex\"></root-widget>\n    </div>",
    }),
    __metadata("design:paramtypes", [JsonSchemaFormService])
], TabComponent);
export { TabComponent };
