// angular2-json-schema-form/src/library/utilities/form-group.functions.js
import { FormArray, FormControl, FormGroup } from '@angular/forms';
import * as _ from 'lodash';
import { forEach, getControlValidators, hasOwn, hasValue, inArray, isArray, isEmpty, isObject, isDefined, isPrimitive, JsonPointer, JsonValidators, toJavaScriptType, toSchemaType, resolveRecursiveReferences } from './index';
/**
 * FormGroup function library:
 *
 * buildFormGroupTemplate:  Builds a FormGroupTemplate from schema
 *
 * buildFormGroup:          Builds an Angular 2 FormGroup from a FormGroupTemplate
 *
 * setRequiredFields:
 *
 * formatFormData:
 *
 * getControl:
 *
 * fixJsonFormOptions:
 *
 * ---- Under construction: ----
 * buildFormGroupTemplateFromLayout: Builds a FormGroupTemplate from a form layout
 */
/**
 * 'buildFormGroupTemplate' function
 *
 * Builds a template for an Angular 2 FormGroup from a JSON Schema.
 *
 * TODO: add support for pattern properties
 * https://spacetelescope.github.io/understanding-json-schema/reference/object.html
 *
 * @param  {any} jsf -
 * @param  {any = null} setValues -
 * @param  {boolean = true} mapArrays -
 * @param  {string = ''} schemaPointer -
 * @param  {string = ''} dataPointer -
 * @param  {any = ''} templatePointer -
 * @return {any} -
 */
export function buildFormGroupTemplate(jsf, setValues, mapArrays, schemaPointer, dataPointer, templatePointer) {
    if (setValues === void 0) { setValues = null; }
    if (mapArrays === void 0) { mapArrays = true; }
    if (schemaPointer === void 0) { schemaPointer = ''; }
    if (dataPointer === void 0) { dataPointer = ''; }
    if (templatePointer === void 0) { templatePointer = ''; }
    var schema = JsonPointer.get(jsf.schema, schemaPointer);
    var useValues = jsf.globalOptions.setSchemaDefaults ?
        mergeValues(JsonPointer.get(schema, '/default'), setValues) : setValues;
    var schemaType = JsonPointer.get(schema, '/type');
    var controlType;
    if (schemaType === 'object' && hasOwn(schema, 'properties')) {
        controlType = 'FormGroup';
    }
    else if (schemaType === 'array' && hasOwn(schema, 'items')) {
        controlType = 'FormArray';
    }
    else if (!schemaType && hasOwn(schema, '$ref')) {
        controlType = '$ref';
    }
    else {
        controlType = 'FormControl';
    }
    if (dataPointer !== '' && !jsf.dataMap.has(dataPointer)) {
        jsf.dataMap.set(dataPointer, new Map);
        jsf.dataMap.get(dataPointer).set('schemaPointer', schemaPointer);
        jsf.dataMap.get(dataPointer).set('schemaType', schema.type);
        if (controlType) {
            jsf.dataMap.get(dataPointer).set('templatePointer', templatePointer);
            jsf.dataMap.get(dataPointer).set('templateType', controlType);
        }
        var genericDataPointer = JsonPointer.toGenericPointer(dataPointer, jsf.arrayMap);
        if (!jsf.dataMap.has(genericDataPointer)) {
            jsf.dataMap.set(genericDataPointer, new Map);
            jsf.dataMap.get(genericDataPointer).set('schemaPointer', schemaPointer);
            jsf.dataMap.get(genericDataPointer).set('schemaType', schema.type);
        }
    }
    var controls;
    var validators = getControlValidators(schema);
    switch (controlType) {
        case 'FormGroup':
            controls = {};
            if (jsf.globalOptions.setSchemaDefaults) {
                useValues = mergeValues(JsonPointer.get(schema, '/properties/default'), useValues);
            }
            forEach(schema.properties, function (item, key) {
                if (key !== 'ui:order') {
                    controls[key] = buildFormGroupTemplate(jsf, JsonPointer.get(useValues, [key]), mapArrays, schemaPointer + '/properties/' + key, dataPointer + '/' + key, templatePointer + '/controls/' + key);
                }
            });
            jsf.globalOptions.fieldsRequired =
                setRequiredFields(schema, controls);
            return { controlType: controlType, controls: controls, validators: validators };
        case 'FormArray':
            var minItems = schema.minItems || 0;
            var maxItems = schema.maxItems || 1000000;
            if (isArray(schema.items)) {
                if (mapArrays && !jsf.arrayMap.get(dataPointer)) {
                    jsf.arrayMap.set(dataPointer, schema.items.length);
                }
                controls = [];
                for (var i = 0, l = schema.items.length; i < l; i++) {
                    if (i >= minItems &&
                        !JsonPointer.has(jsf.templateRefLibrary, [dataPointer + '/' + i])) {
                        jsf.templateRefLibrary[dataPointer + '/' + i] =
                            buildFormGroupTemplate(jsf, null, mapArrays, schemaPointer + '/items/' + i, dataPointer + '/' + i, templatePointer + '/controls/' + i);
                    }
                    if (i < maxItems) {
                        var useValue = isArray(useValues) ? useValues[i] : useValues;
                        controls.push(buildFormGroupTemplate(jsf, useValue, false, schemaPointer + '/items/' + i, dataPointer + '/' + i, templatePointer + '/controls/' + i));
                    }
                }
                if (schema.items.length < maxItems &&
                    hasOwn(schema, 'additionalItems') && isObject(schema.additionalItems)) {
                    var l = Math.max(schema.items.length + 1, isArray(useValues) ? useValues.length : 0);
                    for (var i = schema.items.length; i < l; i++) {
                        var useValue = isArray(useValues) ? useValues[i] : useValues;
                        controls.push(buildFormGroupTemplate(jsf, useValue, false, schemaPointer + '/additionalItems', dataPointer + '/' + i, templatePointer + '/controls/' + i));
                        if (isArray(useValues)) {
                            useValues = null;
                        }
                    }
                    if (!JsonPointer.has(jsf, ['templateRefLibrary', dataPointer + '/-'])) {
                        jsf.templateRefLibrary[dataPointer + '/-'] =
                            buildFormGroupTemplate(jsf, null, mapArrays, schemaPointer + '/additionalItems', dataPointer + '/-', templatePointer + '/controls/-');
                    }
                }
            }
            else {
                if (mapArrays && !jsf.arrayMap.get(dataPointer)) {
                    jsf.arrayMap.set(dataPointer, 0);
                }
                if (!JsonPointer.has(jsf.templateRefLibrary, [dataPointer + '/-'])) {
                    jsf.templateRefLibrary[dataPointer + '/-'] =
                        buildFormGroupTemplate(jsf, null, mapArrays, schemaPointer + '/items', dataPointer + '/-', templatePointer + '/controls/-');
                }
                controls = [];
                if (jsf.globalOptions.setSchemaDefaults) {
                    useValues = mergeValues(JsonPointer.get(schema, '/items/default'), useValues);
                }
                if (isArray(useValues) && useValues.length) {
                    for (var _i = 0, _a = Object.keys(useValues); _i < _a.length; _i++) {
                        var i = _a[_i];
                        controls.push(buildFormGroupTemplate(jsf, useValues[i], false, schemaPointer + '/items', dataPointer + '/' + i, templatePointer + '/controls/' + i));
                    }
                    useValues = null;
                }
            }
            var initialItemCount = Math.max(minItems, JsonPointer.has(schema, '/items/$ref') ? 0 : 1);
            if (controls.length < initialItemCount) {
                for (var i = controls.length, l = initialItemCount; i < l; i++) {
                    controls.push(buildFormGroupTemplate(jsf, useValues, false, schemaPointer + '/items', dataPointer + '/' + i, templatePointer + '/controls/' + i));
                }
            }
            return { controlType: controlType, controls: controls, validators: validators };
        case 'FormControl':
            var value = {
                value: isPrimitive(useValues) ? useValues : null,
                disabled: schema['disabled'] ||
                    JsonPointer.get(schema, '/x-schema-form/disabled') || false
            };
            return { controlType: controlType, value: value, validators: validators };
        case '$ref':
            var schemaRef = JsonPointer.compile(schema.$ref);
            if (!hasOwn(jsf.templateRefLibrary, schemaRef)) {
                // Set to null first to prevent recursive reference from causing endless loop
                jsf.templateRefLibrary[schemaRef] = null;
                var newTemplate = buildFormGroupTemplate(jsf, null, false, schemaRef);
                if (newTemplate) {
                    jsf.templateRefLibrary[schemaRef] = newTemplate;
                }
                else {
                    delete jsf.templateRefLibrary[schemaRef];
                }
            }
            return null;
        default:
            return null;
    }
}
/**
 * 'buildFormGroup' function
 *
 * @param {any} template -
 * @return {AbstractControl}
*/
export function buildFormGroup(template) {
    var validatorFns = [];
    var validatorFn = null;
    if (hasOwn(template, 'validators')) {
        forEach(template.validators, function (parameters, validator) {
            if (typeof JsonValidators[validator] === 'function') {
                validatorFns.push(JsonValidators[validator].apply(null, parameters));
            }
        });
        if (validatorFns.length &&
            inArray(['FormGroup', 'FormArray'], template.controlType)) {
            validatorFn = validatorFns.length > 1 ?
                JsonValidators.compose(validatorFns) : validatorFns[0];
        }
    }
    if (hasOwn(template, 'controlType')) {
        switch (template.controlType) {
            case 'FormGroup':
                var groupControls_1 = {};
                forEach(template.controls, function (controls, key) {
                    var newControl = buildFormGroup(controls);
                    if (newControl) {
                        groupControls_1[key] = newControl;
                    }
                });
                return new FormGroup(groupControls_1, validatorFn);
            case 'FormArray':
                return new FormArray(_.filter(_.map(template.controls, function (controls) { return buildFormGroup(controls); })), validatorFn);
            case 'FormControl':
                return new FormControl(template.value, validatorFns);
        }
    }
    return null;
}
/**
 * 'mergeValues' function
 *
 * @param  {any[]} ...valuesToMerge - Multiple values to merge
 * @return {any} - Merged values
 */
export function mergeValues() {
    var valuesToMerge = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        valuesToMerge[_i] = arguments[_i];
    }
    var mergedValues = null;
    for (var index = 0, length_1 = arguments.length; index < length_1; index++) {
        var currentValue = arguments[index];
        if (!isEmpty(currentValue)) {
            if (typeof currentValue === 'object' &&
                (isEmpty(mergedValues) || typeof mergedValues !== 'object')) {
                if (isArray(currentValue)) {
                    mergedValues = [].concat(currentValue);
                }
                else if (isObject(currentValue)) {
                    mergedValues = Object.assign({}, currentValue);
                }
            }
            else if (typeof currentValue !== 'object') {
                mergedValues = currentValue;
            }
            else if (isObject(mergedValues) && isObject(currentValue)) {
                Object.assign(mergedValues, currentValue);
            }
            else if (isObject(mergedValues) && isArray(currentValue)) {
                var newValues = [];
                for (var _a = 0, currentValue_1 = currentValue; _a < currentValue_1.length; _a++) {
                    var value = currentValue_1[_a];
                    newValues.push(mergeValues(mergedValues, value));
                }
                mergedValues = newValues;
            }
            else if (isArray(mergedValues) && isObject(currentValue)) {
                var newValues = [];
                for (var _b = 0, mergedValues_1 = mergedValues; _b < mergedValues_1.length; _b++) {
                    var value = mergedValues_1[_b];
                    newValues.push(mergeValues(value, currentValue));
                }
                mergedValues = newValues;
            }
            else if (isArray(mergedValues) && isArray(currentValue)) {
                var newValues = [];
                var l = Math.max(mergedValues.length, currentValue.length);
                for (var i = 0; i < l; i++) {
                    if (i < mergedValues.length && i < currentValue.length) {
                        newValues.push(mergeValues(mergedValues[i], currentValue[i]));
                    }
                    else if (i < mergedValues.length) {
                        newValues.push(mergedValues[i]);
                    }
                    else if (i < currentValue.length) {
                        newValues.push(currentValue[i]);
                    }
                }
                mergedValues = newValues;
            }
        }
    }
    return mergedValues;
}
/**
 * 'setRequiredFields' function
 *
 * @param {schema} schema - JSON Schema
 * @param {object} formControlTemplate - Form Control Template object
 * @return {boolean} - true if any fields have been set to required, false if not
 */
export function setRequiredFields(schema, formControlTemplate) {
    var fieldsRequired = false;
    if (hasOwn(schema, 'required') && !isEmpty(schema.required)) {
        fieldsRequired = true;
        var requiredArray = isArray(schema.required) ? schema.required : [schema.required];
        requiredArray = forEach(requiredArray, function (key) { return JsonPointer.set(formControlTemplate, '/' + key + '/validators/required', []); });
    }
    return fieldsRequired;
    // TODO: Add support for patternProperties
    // https://spacetelescope.github.io/understanding-json-schema/reference/object.html
    //   #pattern-properties
}
/**
 * 'formatFormData' function
 *
 * @param {any} formData - Angular 2 FormGroup data object
 * @param  {Map<string, any>} dataMap -
 * @param  {Map<string, string>} recursiveRefMap -
 * @param  {Map<string, number>} arrayMap -
 * @param {boolean = false} fixErrors - if TRUE, tries to fix data
 * @return {any} - formatted data object
 */
export function formatFormData(formData, dataMap, recursiveRefMap, arrayMap, fixErrors) {
    if (fixErrors === void 0) { fixErrors = false; }
    // return formData;
    var formattedData = {};
    JsonPointer.forEachDeep(formData, function (value, dataPointer) {
        if (typeof value !== 'object') {
            var genericPointer = JsonPointer.has(dataMap, [dataPointer, 'schemaType']) ?
                dataPointer :
                resolveRecursiveReferences(dataPointer, recursiveRefMap, arrayMap);
            if (JsonPointer.has(dataMap, [genericPointer, 'schemaType'])) {
                var schemaType = dataMap.get(genericPointer).get('schemaType');
                if (schemaType === 'null') {
                    JsonPointer.set(formattedData, dataPointer, null);
                }
                else if (hasValue(value) &&
                    inArray(schemaType, ['string', 'integer', 'number', 'boolean'])) {
                    var newValue = fixErrors ?
                        toSchemaType(value, schemaType) :
                        toJavaScriptType(value, schemaType);
                    if (isDefined(newValue)) {
                        JsonPointer.set(formattedData, dataPointer, newValue);
                    }
                }
            }
            else {
                console.error('formatFormData error: Schema type not found ' +
                    'for form value at "' + genericPointer + '".');
                console.error(formData);
                console.error(dataMap);
                console.error(recursiveRefMap);
                console.error(arrayMap);
            }
        }
    });
    return formattedData;
}
/**
 * 'getControl' function
 *
 * Uses a JSON Pointer for a data object to retrieve a control from
 * an Angular 2 formGroup or formGroup template. (Note: though a formGroup
 * template is much simpler, its basic structure is idential to a formGroup).
 *
 * If the optional third parameter 'returnGroup' is set to TRUE, the group
 * containing the control is returned, rather than the control itself.
 *
 * @param {FormGroup} formGroup - Angular 2 FormGroup to get value from
 * @param {Pointer} dataPointer - JSON Pointer (string or array)
 * @param {boolean = false} returnGroup - If true, return group containing control
 * @return {group} - Located value (or true or false, if returnError = true)
 */
export function getControl(formGroup, dataPointer, returnGroup) {
    if (returnGroup === void 0) { returnGroup = false; }
    var dataPointerArray = JsonPointer.parse(dataPointer);
    var subGroup = formGroup;
    if (dataPointerArray !== null) {
        var l = dataPointerArray.length - (returnGroup ? 1 : 0);
        for (var i = 0; i < l; ++i) {
            var key = dataPointerArray[i];
            if (subGroup.hasOwnProperty('controls')) {
                subGroup = subGroup.controls;
            }
            if (isArray(subGroup) && (key === '-')) {
                subGroup = subGroup[subGroup.length - 1];
            }
            else if (subGroup.hasOwnProperty(key)) {
                subGroup = subGroup[key];
            }
            else {
                console.error('getControl error: Unable to find "' + key +
                    '" item in FormGroup.');
                console.error(dataPointer);
                console.error(formGroup);
                return;
            }
        }
        return subGroup;
    }
    console.error('getControl error: Invalid JSON Pointer: ' + dataPointer);
}
/**
 * 'fixJsonFormOptions' function
 *
 * Rename JSON Form-style 'options' lists to
 * Angular Schema Form-style 'titleMap' lists.
 *
 * @param  {any} formObject
 * @return {any}
 */
export function fixJsonFormOptions(layout) {
    if (isObject(layout) || isArray(layout)) {
        forEach(layout, function (value, key) {
            if (isObject(value) && hasOwn(value, 'options') && isObject(value.options)) {
                value.titleMap = value.options;
                delete value.options;
            }
        }, 'top-down');
    }
    return layout;
}
