// ng2-kw-gauge/src/app/gauge/gauge.module.js
/**
 * @category   Module
 * @package    com.kiwity.ng2-kw-gauge.gauge
 * @author     Salvador Subarroca (subarroca@gmail.com)
**/
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
// ANGULAR
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
// EXTERNAL
// OWN
import { GaugeComponent } from './gauge.component';
var GaugeModule = (function () {
    function GaugeModule() {
    }
    return GaugeModule;
}());
GaugeModule = __decorate([
    NgModule({
        declarations: [
            GaugeComponent
        ],
        imports: [
            CommonModule
        ],
        providers: [],
        bootstrap: [
            GaugeComponent
        ],
        exports: [
            GaugeComponent
        ]
    })
], GaugeModule);
export { GaugeModule };
